storage: reduce memory consumption for TimestampCache cacheRequests

petermattis opened the pull request:
For ScanRequests which have a truncated response, makeCacheRequest was
maintaining a pointer to the key in the response. This key is allocated
from a buffer which could be significantly larger than the key itself
resulting in significantly more memory usage being referenced than the
TimestampCache than expected.

Fixes #14457

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14514)
<!-- Reviewable:end -->

petermattis committed:
storage: reduce memory consumption for TimestampCache cacheRequests

For ScanRequests which have a truncated response, makeCacheRequest was
maintaining a pointer to the key in the response. This key is allocated
from a buffer which could be significantly larger than the key itself
resulting in significantly more memory usage being referenced than the
TimestampCache than expected.

Fixes #14457

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1605 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_9CXYHHrx4_4WseX2:-Kg_9CXYHHrx4_4WseX3:b-81fr90) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61ab249f9a3f418e860e0a6589fd91f5c8eeab5/pkg/storage/replica.go#L1605)):*
> ```Go
> 					// key was allocated from for the response.
> 					key := resp.Rows[len(resp.Rows)-1].Key
> 					key = key[:len(key):len(key)]
> ```

I think using `len(key)+1` as the capacity of the new array will avoid an allocation in `Next()`. 

---

*[pkg/storage/replica.go, line 1621 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_9Ns24MxOSvlLnhOw:-Kg_9Ns24MxOSvlLnhOx:b-3ihsox) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61ab249f9a3f418e860e0a6589fd91f5c8eeab5/pkg/storage/replica.go#L1621)):*
> ```Go
> 					src := resp.Rows[len(resp.Rows)-1].Key
> 					key := make(roachpb.Key, len(src))
> 					copy(key, src)
> ```

Why are these two implementations different?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1605 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_9CXYHHrx4_4WseX2:-Kg_GV30d-A24ZIXtAz3:bre8woc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61ab249f9a3f418e860e0a6589fd91f5c8eeab5/pkg/storage/replica.go#L1605)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think using `len(key)+1` as the capacity of the new array will avoid an allocation in `Next()`. 
</blockquote></details>

The above wasn't actually allocating a new array but relying on `Next()` to do so. I've changed this to be similar to the implementation in `ReverseScanRequest` which ensures that we allocate the minimum memory necessary.

---

*[pkg/storage/replica.go, line 1621 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_9Ns24MxOSvlLnhOw:-Kg_GgRextexM_rno52H:btbabm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c61ab249f9a3f418e860e0a6589fd91f5c8eeab5/pkg/storage/replica.go#L1621)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why are these two implementations different?
</blockquote></details>

No strong reason. I was being overly clever in forcing `Next` to allocate. I've changed the implementation above to be similar to what is being done here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1603 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_lEeF8R1kWlf5MIkh:-Kg_lEeF8R1kWlf5MIki:b-jqaot8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4763f843f175b52a56ff47bdd7457cdab280a40a/pkg/storage/replica.go#L1603)):*
> ```Go
> 					// perform this truncate if the scan requested a limited number of
> 					// results and we hit that limit.
> 					//
> ```

pedantic nit: this should be an empty line rather than `//`. currently this suggests that the copy has something to do with the preceding comment, but it does not.

---

*[pkg/storage/replica.go, line 1607 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_lUT1LPa2EGNRbvM-:-Kg_lUT1LPa2EGNRbvM0:b-95qkjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4763f843f175b52a56ff47bdd7457cdab280a40a/pkg/storage/replica.go#L1607)):*
> ```Go
> 					// key was allocated from for the response.
> 					src := resp.Rows[len(resp.Rows)-1].Key
> 					key := make(roachpb.Key, len(src), len(src)+1)
> ```

this `len(src)+1` deserves an explicit comment, too.

---

*[pkg/storage/replica.go, line 1619 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_lSVet2RBIVbh3OzR:-Kg_lSVet2RBIVbh3OzS:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/4763f843f175b52a56ff47bdd7457cdab280a40a/pkg/storage/replica.go#L1619)):*
> ```Go
> 					// are returned in reverse order and we truncate the start key of
> 					// the span.
> 					//
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1603 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_lEeF8R1kWlf5MIkh:-KgdoFI2OvVzBWkthLcA:bxvxaa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4763f843f175b52a56ff47bdd7457cdab280a40a/pkg/storage/replica.go#L1603)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

pedantic nit: this should be an empty line rather than `//`. currently this suggests that the copy has something to do with the preceding comment, but it does not.
</blockquote></details>

I'm going to ignore this.

---

*[pkg/storage/replica.go, line 1619 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_lSVet2RBIVbh3OzR:-KgdoI1svbLwEplrTa7l:b-34a71u) ([raw file](https://github.com/cockroachdb/cockroach/blob/4763f843f175b52a56ff47bdd7457cdab280a40a/pkg/storage/replica.go#L1619)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

I'm going to ignore this (too).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14514)*
<!-- Sent from Reviewable.io -->

tamird commented:
=/ OK

On Sat, Apr 1, 2017 at 10:00 AM, Peter Mattis <notifications@github.com>
wrote:

> Review status: all files reviewed at latest revision, 5 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *pkg/storage/replica.go, line 1603 at r2
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_lEeF8R1kWlf5MIkh:-KgdoFI2OvVzBWkthLcA:bxvxaa3>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/4763f843f175b52a56ff47bdd7457cdab280a40a/pkg/storage/replica.go#L1603>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> pedantic nit: this should be an empty line rather than //. currently this
> suggests that the copy has something to do with the preceding comment, but
> it does not.
>
> I'm going to ignore this.
> ------------------------------
>
> *pkg/storage/replica.go, line 1619 at r2
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14514#-Kg_lSVet2RBIVbh3OzR:-KgdoI1svbLwEplrTa7l:b-34a71u>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/4763f843f175b52a56ff47bdd7457cdab280a40a/pkg/storage/replica.go#L1619>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> ditto
>
> I'm going to ignore this (too).
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14514>*
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14514#issuecomment-290921908>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPKlDSWTf3gFs_RDKoCtc3tX49EmJks5rrliYgaJpZM4Mvy8L>
> .
>
